### PR TITLE
macOS: fix linking issues on BigSur

### LINKED ISSUE #6318: Zig on macOS Big Sur
Zig currently doesn't build, nor (for precompiled binaries) run on Big Sur.

### `llvm-config` not found

Use `-DCMAKE_PREFIX_PATH` to have `cmake` look at `/usr/local/opt/llvm`:

```sh
cmake .. \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_PREFIX_PATH=$(brew --prefix llvm)
```

### `cmake` cannot find the compiler ID

Update: fixed in recent Zig versions.

### `libSystem` not found

__Workaround:__ presumably fixed in LLVM12, but until then, use the system linker hack and set the sysroot.

The system linker doesn't support `'-error-limit`, so don't add this option when using it.

The sysroot can be set by adding `-syslibroot` following by the path returned by the `xcrun --show-sdk-path` command.

Super ugly patch (or use [that branch](https://github.com/jedisct1/zig/tree/bigsur) which is constantly kept in sync with the main branch):

```diff
diff --git a/src/link/MachO.zig b/src/link/MachO.zig
index 7344a698f..54508718c 100644
--- a/src/link/MachO.zig
+++ b/src/link/MachO.zig
@@ -397,6 +397,19 @@ pub fn flushModule(self: *MachO, comp: *Compilation) !void {
     assert(!self.libsystem_cmd_dirty);
 }
 
+fn getSDKPath(self: *MachO) ![]const u8 {
+    const argv = [_][]const u8{ "/usr/bin/xcrun", "--show-sdk-path" };
+    const result = try std.ChildProcess.exec(.{ .allocator = self.base.allocator, .argv = &argv });
+    if (result.stderr.len != 0) {
+        std.log.err("unexpected xcrun stderr: {}", .{result.stderr});
+    }
+    self.base.allocator.free(result.stderr);
+    if (result.term.Exited != 0) {
+        return error.XcrunReportedFailure;
+    }
+    return std.mem.trimRight(u8, result.stdout, "\r\n");
+}
+
 fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
     const tracy = trace(@src());
     defer tracy.end();
@@ -529,14 +542,16 @@ fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
             try fs.cwd().copyFile(the_object_path, fs.cwd(), full_out_path, .{});
         }
     } else {
-        // Create an LLD command line and invoke it.
+        // Create an LD command line and invoke it.
         var argv = std.ArrayList([]const u8).init(self.base.allocator);
         defer argv.deinit();
-        // Even though we're calling LLD as a library it thinks the first argument is its own exe name.
-        try argv.append("lld");
 
-        try argv.append("-error-limit");
-        try argv.append("0");
+        try argv.append("/usr/bin/ld");
+
+        const sdk_path = try self.getSDKPath();
+        defer self.base.allocator.free(sdk_path);
+        try argv.append("-syslibroot");
+        try argv.append(try self.getSDKPath());
 
         try argv.append("-demangle");
 
@@ -562,6 +577,7 @@ fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
         }
 
         if (is_dyn_lib) {
+            try argv.append("-headerpad_max_install_names");
             try argv.append("-dylib");
 
             if (self.base.options.version) |ver| {
@@ -713,28 +729,19 @@ fn linkWithLLD(self: *MachO, comp: *Compilation) !void {
             .data = std.ArrayList(u8).init(self.base.allocator),
         };
         defer stdout_context.data.deinit();
-        const llvm = @import("../llvm.zig");
-        const ok = llvm.Link(
-            .MachO,
-            new_argv.ptr,
-            new_argv.len,
-            append_diagnostic,
-            @ptrToInt(&stdout_context),
-            @ptrToInt(&stderr_context),
-        );
-        if (stderr_context.oom or stdout_context.oom) return error.OutOfMemory;
-        if (stdout_context.data.items.len != 0) {
-            std.log.warn("unexpected LLD stdout: {}", .{stdout_context.data.items});
+
+        const result = try std.ChildProcess.exec(.{ .allocator = self.base.allocator, .argv = argv.items });
+        if (result.stdout.len != 0) {
+            std.log.warn("unexpected LD stdout: {}", .{result.stdout});
         }
-        if (!ok) {
-            // TODO parse this output and surface with the Compilation API rather than
-            // directly outputting to stderr here.
-            std.debug.print("{}", .{stderr_context.data.items});
+        if (result.stderr.len != 0) {
+            std.log.warn("unexpected LD stderr: {}", .{result.stderr});
+        }
+        self.base.allocator.free(result.stderr);
+        self.base.allocator.free(result.stdout);
+        if (result.term.Exited != 0) {
             return error.LLDReportedFailure;
         }
-        if (stderr_context.data.items.len != 0) {
-            std.log.warn("unexpected LLD stderr: {}", .{stderr_context.data.items});
-        }
     }
 
     if (!self.base.options.disable_lld_caching) {
diff --git a/src/main.zig b/src/main.zig
index 8162bc46b..6711f802c 100644
--- a/src/main.zig
+++ b/src/main.zig
@@ -1707,7 +1707,7 @@ fn buildOutputType(
         }
     };
 
-    try updateModule(gpa, comp, zir_out_path, hook);
+    updateModule(gpa, comp, zir_out_path, hook) catch std.process.exit(1);
 
     if (build_options.is_stage1 and comp.stage1_lock != null and watch) {
         warn("--watch is not recommended with the stage1 backend; it leaks memory and is not capable of incremental compilation", .{});
```

### Can't link `libxml2.tbd`

`findLLVM` wants to include `libxml2` to linked libraries but does it by adding the `-llibxml2.tbd` option instead of `-lxml2`. 

This is a bug in LLVM. When installed via Homebrew, LLVM 11 now includes a patch that fixes it.

---

Eventually, `zig` perfectly works on Big Sur.

Sorry for that enumeration of hacks rather sending a proper PR. Hope it can still help people who already upgraded to Big Sur.

### PR DESCRIPTION
**Pull request by kubkon**
This commit fixes linking issue on macOS 11 BigSur by appending
a prefix path to all lib and framework search paths known as
`-syslibroot`.

The reason this is needed is that in macOS 11, the system libraries
and frameworks are no longer readily available in the filesystem.
Instead, the new macOS ships with a built-in dynamic linker cache
of all system-provided libraries, and hence, when linking with either
`lld.ld64` or `ld64`, it is required to pass in `-syslibroot [dir]`.
The latter can usually be obtained by invoking `xcrun --show-sdk-path`.
With this commit, Zig will do this automatically when compiling natively
on macOS. However, it also provides a flag `-syslibroot` which can be
used to overwrite the automtically populated value.

To summarise, with this change, the user of Zig is not required to
generate and append their own syslibroot path. Standard invocations
such as `zig build-exe hello.zig` or `zig build` for projects will
work out of the box. The only missing bit is `zig cc` and `zig c++`
since the addition of the `-syslibroot` option would be a mismatch
between the values provided by `clang` itself and Zig's wrapper.

Closes #6318 

cc @jedisct1

### CODE REVIEW

File: lib/std/build.zig [hunk 1]
```diff
@@ -1841,6 +1842,10 @@ pub const LibExeObjStep = struct {
         self.lib_paths.append(self.builder.dupe(path)) catch unreachable;
     }
 
+    pub fn addSyslibroot(self: *LibExeObjStep, path: []const u8) void {
+        self.syslibroot = path;
+    }
+
```
Comment (andrewrk) [inline lib/std/build.zig:1848]: counter proposal: omit this method and add doc comments to the field indicating it is OK for the user to set the field directly.

File: lib/std/build.zig [hunk 2]
```diff
@@ -2271,6 +2283,18 @@ pub const LibExeObjStep = struct {
         }
 
         if (self.target.isDarwin()) {
+            if (self.syslibroot) |path| {
+                try zig_args.append("-syslibroot");
+                try zig_args.append(path);
+            } else {
+                if (self.target.isNative()) {
+                    const syslibroot = try std.zig.system.getSDKPath(builder.allocator);
```
Comment (andrewrk) [inline lib/std/build.zig:2291]: We shouldn't need this call in std.build because we expect this to work:

```
zig build-exe something_that_requires_libsysroot.zig
```

This means if the target OS is native and we omit the -syslibroot parameter then the compiler will figure it out (probably by doing getSDKPath itself).

File: lib/std/zig/system/macos.zig [hunk 3]
```diff
@@ -452,3 +454,25 @@ test "version_from_build" {
...
     }
 }
+
+/// Detect SDK path on Darwin.
+/// Calls `xcrun --show-sdk-path` which result can be used to specify
+/// `-syslibroot` param of the linker.
+/// The caller needs to free the resulting path slice.
+pub fn getSDKPath(allocator: *mem.Allocator) ![]u8 {
+    assert(std.Target.current.isDarwin());
+    const argv = &[_][]const u8{ "/usr/bin/xcrun", "--show-sdk-path" };
```
Comment (andrewrk) [inline lib/std/zig/system/macos.zig:464]: Is this intentionally ignoring PATH with respect to `xcrun`? (not ignoring PATH would look like chopping off the "/usr/bin/")

File: src/main.zig [hunk 4]
```diff
@@ -315,8 +315,9 @@ const usage_build_generic =
     \\  --subsystem [subsystem]        (windows) /SUBSYSTEM:<subsystem> to the linker\n"
     \\  --stack [size]                 Override default stack size
     \\  --image-base [addr]            Set base address for executable image
+    \\  -syslibroot [dir]              (darwin) prepend a prefix to all search paths
     \\  -framework [name]              (darwin) link against framework
-    \\  -F[dir]                        (darwin) add search path for frameworks
+    \\  -F [dir]                       (darwin) add search path for frameworks
```
Comment (andrewrk) [inline src/main.zig:320]: This change is fine, but just FYI we do actually support both

File: src/main.zig [hunk 5]
```diff
@@ -1380,6 +1386,12 @@ fn buildOutputType(
         }
     }
 
+    if (cross_target.isNativeOs() and std.Target.current.isDarwin()) {
+        if (syslibroot == null) {
+            syslibroot = try std.zig.system.getSDKPath(arena);
+        }
+    }
```
Comment (andrewrk) [inline src/main.zig:1393]: I think we should move this code into Compilation.zig so that we can do it only in the path where `use_lld` is `true`. No sense in wasting precious milliseconds executing `xcrun` when your self-hosted linker code does not need this information.

https://github.com/ziglang/zig/blob/909aae8153e7ac71197bc9f864e4ceb2793317f2/src/Compilation.zig#L443

After we have this flag you could put this logic there. Also if you gate it behind `build_options.have_llvm` which is a comptime value, then we won't even have the code bloat of the ability to execute `xcrun` for non-llvm-enabled builds.

File: src/main.zig [hunk 6]
```diff
@@ -2200,6 +2214,11 @@ pub fn cmdBuild(gpa: *Allocator, arena: *Allocator, args: []const []const u8) !v
                         override_global_cache_dir = args[i];
                         try child_argv.appendSlice(&[_][]const u8{ arg, args[i] });
                         continue;
+                    } else if (mem.eql(u8, arg, "--syslibroot")) {
+                        if (i + 1 >= args.len) fatal("expected argument after '{}'", .{arg});
+                        i += 1;
+                        syslibroot = args[i];
+                        continue;
```
Comment (andrewrk) [inline src/main.zig:2221]: The CLI options here are only options that could affect the compilation of the build runner itself. If your goal here is to add a `--syslibroot` general option that is always available to `zig build` (which I'm not actually sure is appropriate) then that would go in std/special/build_runner.zig.

I understand that this value could, in fact, influence the build executable, however, we always want to run the build exe for the native target, so I think it would be a simplification here to assume that, and delete this logic. Combining this with my other suggestion about moving the logic to Compilation, and the changes to `cmdBuild` can simply be deleted.

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: lib/std/build.zig (hunk 1 above)
Comment (kubkon) [reply lib/std/build.zig:1848]: I like that, thanks!

File: src/main.zig (hunk 4 above)
Comment (kubkon) [reply src/main.zig:320]: Oh ok. I'm happy to revert!

File: src/main.zig (hunk 5 above)
Comment (kubkon) [reply src/main.zig:1393]: Nice! This is an excellent suggestion!

File: src/Compilation.zig [hunk 7]
```diff
@@ -472,6 +472,11 @@ pub fn create(gpa: *Allocator, options: InitOptions) !*Compilation {
             break :blk false;
         };
 
+        const syslibroot = if (use_lld and options.is_native_os and options.target.isDarwin()) blk: {
```
Comment (andrewrk) [inline src/Compilation.zig:475]: `use_lld` is always false when build_options.have_llvm is false. So if we prefix this with it, we prevent getSDKPath from even being compiled when LLD is a build dependency.

```suggestion
        const syslibroot = if (build_options.have_llvm and use_lld and options.is_native_os and options.target.isDarwin()) blk: {
```

**Review by andrewrk — APPROVED** [review]
Looks good, one last suggestion, but I think it's already mergable :+1:

File: src/Compilation.zig (hunk 7 above)
Comment (andrewrk) [reply src/Compilation.zig]: Also might make sense to put `std.Target.current.isDarwin()` in there since that must be true for `options.is_native_os and options.target.isDarwin()` to be true.
